File: client/src/Envelope.kt
package dev.yellobytes.ghostbe.client

import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.boolean
import kotlinx.serialization.json.decodeFromJsonElement
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

private val json = Json { ignoreUnknownKeys = true }

@Serializable
data class RequestEnvelope(
    val method: String,
    val url: String,
    val headers: Map<String, String>,
    val body: String? = null
)

fun RequestEnvelope.toJson(): String = json.encodeToString(this)

@Serializable
sealed interface ResponseEnvelope {
    @Serializable
    data class Mock(
        val intercept: Boolean = true,
        val status: Int,
        val headers: Map<String, String>,
        val body: String
    ) : ResponseEnvelope

    @Serializable
    data class Passthrough(val intercept: Boolean = false) : ResponseEnvelope

    companion object
}

fun ResponseEnvelope.Companion.fromJson(text: String): ResponseEnvelope {
    val element = json.parseToJsonElement(text).jsonObject
    val intercept = element["intercept"]?.jsonPrimitive?.boolean ?: false
    return if (intercept) {
        json.decodeFromJsonElement<ResponseEnvelope.Mock>(element)
    } else {
        ResponseEnvelope.Passthrough()
    }
}

File: client-flutter/ghost_be/lib/src/envelope.dart
class RequestEnvelope {
  final String method;
  final String url;
  final Map<String, String> headers;
  final String? body;

  RequestEnvelope({
    required this.method,
    required this.url,
    required this.headers,
    this.body,
  });

  Map<String, dynamic> toJson() => {
        'method': method,
        'url': url,
        'headers': headers,
        'body': body,
      };
}

sealed class ResponseEnvelope {
  factory ResponseEnvelope.fromJson(Map<String, dynamic> json) {
    final intercept = json['intercept'] as bool? ?? false;
    if (intercept) {
      return MockResponseEnvelope(
        status: json['status'] as int,
        headers: Map<String, String>.from(json['headers'] as Map),
        body: json['body'] as String,
      );
    }
    return PassthroughResponseEnvelope();
  }
}

class MockResponseEnvelope implements ResponseEnvelope {
  final int status;
  final Map<String, String> headers;
  final String body;

  MockResponseEnvelope({
    required this.status,
    required this.headers,
    required this.body,
  });
}

class PassthroughResponseEnvelope implements ResponseEnvelope {}
